plonky3/src/prover/mod.rs

```diff
@@ -0,0 +1,203 @@
+    pub fn write_setup(&self, output: &mut dyn std::io::Write) {
+        serde_json::to_writer(output, &self.setup.values).unwrap();
+    }
+
+    pub fn write_vkey(&self, output: &mut dyn std::io::Write) {
+        let vk = self.verification_key();
+        serde_json::to_writer(output, &vk.values).unwrap();
+    }
+
+    pub fn verification_key(&self) -> &Constants {
+        &self.setup
+    }
+
+    pub fn new(
+        analyzed: &'a Analyzed<T>,
+        fixed: &'a [(String, Vec<T>)],
+        constants: Option<&mut dyn std::io::Read>,
+    ) -> Result<Self, std::io::Error> {
+        let constants = constants
+            .map(serde_json::from_reader)
+            .transpose()?
+            .unwrap_or_else(generate_setup);
+
+        Ok(Self {
+            analyzed,
+            fixed,
+            setup: Constants::new(constants),
+            vkey: None,
+        })
+    }
+
+    pub fn add_verification_key(&mut self, vkey: &mut dyn Read) {
+        self.vkey = Some(Constants::new(serde_json::de::from_reader(vkey).unwrap()));
+    }
+}
+
+impl<'a, T: FieldElement> Plonky3Prover<'a, T> {
+    pub fn prove_ast(
+        &self,
+        witness: &[(String, Vec<T>)],
+        witgen_callback: WitgenCallback<T>,
+    ) -> Result<Vec<u8>, String> {
+        assert_eq!(T::known_field(), Some(KnownField::GoldilocksField));
+
+        let circuit = PowdrCircuit::new(self.analyzed, self.fixed)
+            .with_witgen_callback(witgen_callback)
+            .with_witness(witness);
+        let publics = circuit.instance_column();
+
+        let trace = circuit.preprocessed_trace().unwrap();
+
+        let (config, perm) = self.setup.to_config_and_perm(self.analyzed.degree());
+
+        let mut challenger = Challenger::new(perm.clone());
+
+        let proof = prove(&config, &circuit, &mut challenger, trace, &publics);
+
+        let mut challenger = Challenger::new(perm);
+
+        verify(&config, &circuit, &mut challenger, &proof, &publics).unwrap();
+        Ok(serde_json::to_vec(&proof).unwrap())
+    }
+
+    pub fn verify(&self, proof: &[u8], instances: &[Vec<T>]) -> Result<(), String> {
+        let proof: Proof<_> = serde_json::from_slice(proof)
+            .map_err(|e| format!("Failed to deserialize proof: {e}"))?;
+        let publics = instances
+            .iter()
+            .flatten()
+            .map(|v| cast_to_goldilocks(*v))
+            .collect();
+
+        let (config, perm) = self
+            .vkey
+            .as_ref()
+            .unwrap()
+            .to_config_and_perm(self.analyzed.degree());
+
+        let mut challenger = Challenger::new(perm);
+
+        verify(
+            &config,
+            &PowdrCircuit::new(self.analyzed, self.fixed),
+            &mut challenger,
+            &proof,
+            &publics,
+        )
+        .map_err(|e| format!("Failed to verify proof: {e:?}"))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use powdr_number::GoldilocksField;
+    use powdr_pipeline::Pipeline;
```

Comment:
Not sure we should have Pipeline as a dependency here, even as a dev-dep. Maybe better to keep these tests together with the others?

Reply:
Would this be fixed by moving plonky3 into backend?